utils/utils.py

```diff
@@ -71,55 +72,188 @@ def train_validate_test_normal(
+    else:
+        model = model_wrapper
+
+    x_atomfeature = []
+    num_nodes = len(test_loader.dataset[0].x)
+    for data in test_loader.dataset:
+        x_atomfeature.append(data.x)
+    if plot_init_sol:  # visualizing of initial conditions
+        test_rmse = test(test_loader, model, config["Architecture"]["output_dim"])
+        true_values = test_rmse[3]
+        predicted_values = test_rmse[4]
+        for ihead in range(model.num_heads):
+            visualizer = Visualizer(model_with_config_name)
+            visualizer.add_test_values(
+                true_values=true_values[ihead], predicted_values=predicted_values[ihead]
+            )
+            if model.head_dims[ihead] // num_nodes == 3:  # magnetic moments
+                visualizer.create_scatter_plot_atoms_vec(
+                    config["Target_dataset"]["output_names"][ihead], x_atomfeature, -1
+                )
+            else:
+                visualizer.create_scatter_plot_atoms(
+                    config["Target_dataset"]["output_names"][ihead], x_atomfeature, -1
+                )
+                visualizer.create_error_histogram_plot_atoms(
+                    config["Target_dataset"]["output_names"][ihead], x_atomfeature, -1
+                )
+
     for epoch in range(0, num_epoch):
-        train_mae = train(train_loader, model, optimizer, config["output_dim"])
-        val_mae = validate(val_loader, model, config["output_dim"])
-        test_rmse = test(test_loader, model, config["output_dim"])
+        train_mae, train_taskserr, train_taskserr_nodes = train(
+            train_loader, model, optimizer, config["Architecture"]["output_dim"]
+        )
+        val_mae, val_taskserr, val_taskserr_nodes = validate(
+            val_loader, model, config["Architecture"]["output_dim"]
+        )
+        test_rmse = test(test_loader, model, config["Architecture"]["output_dim"])
         scheduler.step(val_mae)
         writer.add_scalar("train error", train_mae, epoch)
         writer.add_scalar("validate error", val_mae, epoch)
         writer.add_scalar("test error", test_rmse[0], epoch)
+        for ivar in range(model.num_heads):
+            writer.add_scalar(
+                "train error of task" + str(ivar), train_taskserr[ivar], epoch
+            )
 
         print(
             f"Epoch: {epoch:02d}, Train MAE: {train_mae:.8f}, Val MAE: {val_mae:.8f}, "
             f"Test RMSE: {test_rmse[0]:.8f}"
         )
+        print("Tasks MAE:", train_taskserr)
+
+        trainlib.append(train_mae)
+        vallib.append(val_mae)
+        testlib.append(test_rmse[0])
+        tasklib.append(train_taskserr)
+        tasklib_vali.append(val_taskserr)
+        tasklib_test.append(test_rmse[1])
+
+        tasklib_nodes.append(train_taskserr_nodes)
+        tasklib_vali_nodes.append(val_taskserr_nodes)
+        tasklib_test_nodes.append(test_rmse[2])
+
+        ###tracking the solution evolving with training
+        if plot_hist_sol:
```

Comment:
the plot of the solution should be a separate function. The visualization techniques should be part of the training. 
In fact, I would be nice if the training function returned the history of training-valdiation-test-losses, and then these were used by another function